src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -369,5 +382,41 @@ internal static async Task SetAppOfflineState(string rootPath, bool offline)
                 }, maxRetries: 3, retryInterval: TimeSpan.FromSeconds(1));
             }
         }
+
+        private class SuspendRestartRequest : IDisposable
+        {
+            private FileMonitoringService _fileMonitoringService;
+
+            public SuspendRestartRequest(FileMonitoringService fileMonitoringService)
+            {
+                _fileMonitoringService = fileMonitoringService;
+                if (Interlocked.Increment(ref _fileMonitoringService._suspensionRequestsCount) == 1)
+                {
+                    _fileMonitoringService._typedLogger.LogDebug("Restart is suspended");
+                }
+            }
+
+            public void Dispose()
+            {
+                if (Interlocked.Decrement(ref _fileMonitoringService._suspensionRequestsCount) == 0)
+                {
+                    _fileMonitoringService._typedLogger.LogDebug("Restart is resumed");
+                    if (Interlocked.Read(ref _fileMonitoringService._restartRequested) == 1)
+                    {
+                        _fileMonitoringService._typedLogger.LogDebug("Initiate force restart as it was triggerd during the suspension");
+                        // Blocking call as we can't use async call in Dispose method
+                        _fileMonitoringService._scriptHostManager.RestartHostAsync().GetAwaiter().GetResult();
```

Comment:
This differs a bit from the behavior we had discussed. Would the caller expect a restart to happen if something would have triggered it while it was suspended? I think this would come as a surprise and would also limit the caller's choice as to when/if a restart should happen. 

I think we should just resume watching and not restart. It would be up to the caller to do so.

Reply:
this make sense so caller code will be changed to:
```
using(fileMonitoringService?.SuspendRestart()) // increment suspensionRequestCount
{
..
} // decrement suspensionRequestCount
await fileMonitoringService?.TryRestartAsyc()
```

`TryRestartAsync` will restart only if `suspensionRequestCount== 0`